src/Conversion/ONNXToTOSA/Tensor/Concat.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Concat.cpp - Concat Op --------------------===//
+//
+// Copyright (c) 2023 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX ConcatOp to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXConcatLoweringToTOSA : public OpConversionPattern<ONNXConcatOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  using OpAdaptor = typename ONNXConcatOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXConcatOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
+    ValueRange inputs = adaptor.inputs();
+    int64_t axis = adaptor.axis();
+    auto resultType = op.getResult().getType();
+
+    for (const auto &input : inputs) {
+      if (!onnx_mlir::isRankedShapedType(input.getType()))
+        return rewriter.notifyMatchFailure(
+            op, "inputs are not ranked shaped tensors");
+    }
+    int64_t inputRank = onnx_mlir::getRank(inputs[0].getType());
+
+    // onnx allows values beetween [-r, r-1] where r is the rank.
+    if (axis < 0)
+      axis += inputRank;
+
+    // Check if axis is in correct range.
+    if (axis < 0 || axis >= inputRank)
+      return rewriter.notifyMatchFailure(
+          op, "axis attribute not in correct range");
```

Comment:
Have we this in more places? If so, it would be good to pull it into common code.

Reply:
The check is new, but now that you say it should also be used for the other places where we modify axis. I will add a helper function